### ISSUE TITLE
Automated Updates

### ISSUE THREAD

**Issue by gardener-robot-ci-1** [gardener/gardener#33]
**Issue by [kubernetes-jenkins](https://github.com/kubernetes-jenkins)**
_Wednesday Jul 19, 2017 at 18:45 GMT_
_Originally opened as https://git.removed/kubernetes-attic/garden-operator/issues/53_

----

# Stories
* As user I want the latest version with all security patches and bug fixes, so that my cluster is safe and sound.
* As provider I can update the OS, so that I can deploy important security patches and bug fixes quickly.
* As provider I can update Kubernetes and its components, so that I can deploy important security patches and bug fixes quickly.

# Motivation
See above.

# Acceptance Criteria
- [ ] Changes to the OS can be rolled through all self-controlled landscapes
- [ ] Changes to Kubernetes and its components can be rolled through all self-controlled landscapes
- [ ] CI process that keeps track of patches, automatically pre-validates them, and prepares a PR for review

:information_source: Migrated from Jira issue [KUBE-16](https://jtrack.removed/browse/KUBE-16)

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Friday Oct 06, 2017 at 07:58 GMT_

----

Our pipeline should include the following:

On every dev codeline PR/update:
* Formal checks like referenced issue
* Lint checks
* Unit tests (mocking everything outside the unit/module)
* Acceptance tests (mocking everything outside the component/GitHub project)

On every release codeline PR/update:
* Deployment into a Garden DEV cluster for testing purposes, not watching all namespaces, but only an integration-test namespace:
  * Create, consume, and delete a cluster on all infrastructure in parallel
  * Take an existing cluster and update it

Testing matrix:
* One dimension is the IaaS, i.e. AWS, Azure, GCP, and OpenStack
* Another dimension is creation respectively update (from the latest operator version for that particular Kubernetes version) of the various supported versions (maybe we can limit creation to fewer versions than update)

| | AWS | Azure | GCP | OpenStack |
|---|---|---|---|---|
Creation of new v1.9 | ☀️ | ❌ | ☀️ | ❌ |
Creation of new v1.8 | ☀️ | ☀️ | ☀️ | ☀️ |
Update to latest v1.9 | ☀️ | ☀️ | ☀️ | ❌ |
Update to latest v1.8 | ☀️ | ❌ | ☀️ | ❌ |
Update to latest v1.7 | ❌ | ☀️ | ☀️ | ☀️ |
Update to latest v1.6 | ☀️ | ☀️ | ☀️ | ❌ |

After it gets deployed to production:
* Start first with Canary updates (single clusters before going crazy on all clusters); if we have designated namespaces, start with the less critical ones, maybe internal/our, but not the seed clusters, of course, which should come last
* Scale out to all namespaces and do that in parallel, but still limited (no more than a configurable percentage, e.g. 20%)

**Comment by vlerenc** [comment]
Duplicate of https://github.com/gardener/gardener/issues/14.
